File: docs/src/testers/file-storage/descriptive_filenames.py
# @sniptest filename=descriptive_filenames.py
# @sniptest show=19-26
from pathlib import Path

from notte_sdk.endpoints.files import RemoteFileStorage

# Record each successful upload before any subsequent request can fail.
owned_uploads = []
original_upload = RemoteFileStorage.upload


def tracked_upload(self, *args, **kwargs):
    uploaded = original_upload(self, *args, **kwargs)
    owned_uploads.append((self, uploaded.id))
    return uploaded


RemoteFileStorage.upload = tracked_upload
primary_error = None
try:
    from datetime import datetime

    from notte_sdk import NotteClient

    client = NotteClient()
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    with client.Session() as session:
        session.storage.upload("report.pdf", upload_file_name=f"report_{timestamp}.pdf")
        status = session.status()

    files = session.storage.list(source="user_upload").files
    assert sorted(file.filename for file in files) == [f"report_{timestamp}.pdf"]
    for file in files:
        downloaded = session.storage.download(file.id, local_dir="./verified")
        assert Path(downloaded).read_bytes() == Path("report.pdf").read_bytes()
except BaseException as error:
    primary_error = error
    raise
finally:
    RemoteFileStorage.upload = original_upload
    cleanup_errors = []
    for storage, file_id in owned_uploads:
        try:
            storage.delete(file_id)
        except Exception as error:
            cleanup_errors.append(error)
    if cleanup_errors:
        if primary_error is not None:
            raise BaseExceptionGroup("Example and cleanup failed", [primary_error, *cleanup_errors]) from None
        raise ExceptionGroup("Failed to delete uploaded example files", cleanup_errors)

File: docs/src/testers/file-storage/descriptive_filenames.ts
// @sniptest filename=descriptive_filenames.ts
// @sniptest show=19-29
import assert from 'node:assert/strict';
import { readFile } from 'node:fs/promises';
import { NotteClient, RemoteFileStorage, sessionStatus, type SessionResponse } from 'notte-sdk';

// Record each successful upload before any subsequent request can fail.
const ownedUploads: Array<{ storage: RemoteFileStorage; id: string }> = [];
const originalUpload = RemoteFileStorage.prototype.upload;
RemoteFileStorage.prototype.upload = async function (...args) {
  const uploaded = await originalUpload.apply(this, args);
  ownedUploads.push({ storage: this, id: uploaded.id });
  status = (await sessionStatus({
    client: new NotteClient().getClient(), path: { session_id: this.sessionId }, throwOnError: true,
  })).data;
  return uploaded;
};
let status: SessionResponse;
const primaryErrors: unknown[] = [];
try {
  const { NotteClient } = await import('notte-sdk');

  const client = new NotteClient();
  const now = new Date();
  const date = [now.getFullYear(), now.getMonth() + 1, now.getDate()].map(n => String(n).padStart(2, '0')).join('');
  const time = [now.getHours(), now.getMinutes(), now.getSeconds()].map(n => String(n).padStart(2, '0')).join('');
  const timestamp = `${date}_${time}`;
  const storage = client.FileStorage();
  await client.Session({ storage }).use(async session => {
    await storage.upload('report.pdf', `report_${timestamp}.pdf`);
  });

  const { files } = await storage.list({ source: 'user_upload' });
  assert.deepEqual(files.map(file => file.filename).sort(), [`report_${timestamp}.pdf`]);
  for (const file of files) {
    assert.deepEqual(Buffer.from(await (await storage.downloadBlob(file.id)).arrayBuffer()), await readFile('report.pdf'));
  }
} catch (error) {
  primaryErrors.push(error);
  throw error;
} finally {
  RemoteFileStorage.prototype.upload = originalUpload;
  const cleanup = await Promise.allSettled(ownedUploads.map(file => file.storage.delete(file.id)));
  const errors = cleanup.filter(result => result.status === 'rejected').map(result => result.reason);
  if (errors.length) throw new AggregateError([...primaryErrors, ...errors], 'Example file cleanup failed');
}
export { status };
